src/Parser/Druid/Feral/Modules/Features/FerociousBiteEnergy.js

```diff
@@ -0,0 +1,160 @@
+import React from 'react';
+
+import Analyzer from 'Parser/Core/Analyzer';
+import Combatants from 'Parser/Core/Modules/Combatants';
+
+import SPELLS from 'common/SPELLS';
+import RESOURCE_TYPES from 'common/RESOURCE_TYPES';
+
+import SpellLink from 'common/SpellLink';
+import Wrapper from 'common/Wrapper';
+import { formatNumber } from 'common/format';
+
+const ENERGY_MIN_USED_BY_BITE = 25;
+const ENERGY_FOR_FULL_DAMAGE_BITE = 50;
+const MAX_DAMAGE_BONUS_FROM_ENERGY = 1.0;
+const LEEWAY_BETWEEN_CAST_AND_DAMAGE = 100; // in thousandths of a second
+
+/**
+ * Although Ferocious Bite costs 25 energy, it does up to double damage if the character has more.
+ * It's recommended that feral druids use Bite when at 50 energy or higher.
+ * An exception to this is when the bonus from 4-piece T21 is active, which makes Bite cost no
+ * energy and ignore the current energy level.
+ */
+class FerociousBiteEnergy extends Analyzer {
+  static dependencies = {
+    combatants: Combatants,
+  };
+  
+  biteCount = 0;
+  freeBiteCount = 0;
+  lowEnergyBiteCount = 0;
+  lostDamageTotal = 0;
+  energySpentOnBiteTotal = 0;
+  
+  biteDamageEvents = [];
+  lowEnergyBiteCastEvents = [];
+  
+  on_byPlayer_damage(event) {
+    if (event.ability.guid !== SPELLS.FEROCIOUS_BITE.id) {
+      return;
+    }
+
+    // Damage event only provides a feral druid's mana in classResources, not their energy.
+    // So store damage and link it with energy information from cast event after parsing is complete.
+    this.biteDamageEvents.push(event);
+  }
+
+  on_byPlayer_cast(event) {
+    if (event.ability.guid !== SPELLS.FEROCIOUS_BITE.id) {
+      return;
+    }
+    this.biteCount++;
+
+    const resource = event.classResources[0];
+    if (resource.type !== RESOURCE_TYPES.ENERGY.id || !resource.cost) {
+      // Ferocious Bite didn't consume energy, and so ignores the character's energy level.
+      this.freeBiteCount++;
+      return;
+    }
+
+    // 'amount' is energy before the ability uses it.
+    if (resource.amount < ENERGY_FOR_FULL_DAMAGE_BITE) {
+      this.energySpentOnBiteTotal += resource.amount;
+      this.lowEnergyBiteCount++;
+      this.lowEnergyBiteCastEvents.push(event);
+    }
+    else {
+      this.energySpentOnBiteTotal += ENERGY_FOR_FULL_DAMAGE_BITE;
+    }
+  }
+
+  on_finished() {
+    this.lowEnergyBiteCastEvents.forEach((castEvent) => {
```

Comment:
Is there a reason you're searching through the bite events after the fact rather than doing it during parsing? Can bite damage events occur before casts? If not, then you can simply:

1. record the energy spent on the bite in `on_byPlayer_cast`
2. calculate the damage lost using that number in `on_byPlayer_damage`

This avoids doing the extra searching at the end. Admittedly, it isn't *expensive* but it is unnecessary and unconventional.